### ISSUE TITLE
How to change the number of channels

### ISSUE THREAD

**Issue by LittleWat** [rbgirshick/py-faster-rcnn#239]
Hi, I would like to change the 3 channel (RGB) to 4 or more channels.

I mainly changed the codes where "cv2.imread" function are called to accept 4channel picture.
but error occured as below.

F0703 18:00:10.648262 34300 base_conv_layer.cpp:192] Check failed: bottom[0]->shape(channel_axis_) == channels_ (4 vs. 3) Input size incompatible with convolution kernel.

Is it necessary to change the .cpp  files?

Is there any easy way to change the number of channels.

Sincerely

**Comment by GBJim** [comment]
Hi @LittleWat 
Here is a related discussion:  [#232](https://github.com/rbgirshick/py-faster-rcnn/issues/232)

**Comment by LittleWat** [comment]
@GBJim 
Thank you for telling related discussion.

**Comment by LittleWat** [comment]
I could solved this issue
#  >shape(channel_axis_) == channels_ (4 vs. 3) Input size incompatible with convolution kernel.

by changing the first layer name from "conv1_1" to "my_conv1_1" in train.prototxt.

but after I change that name and train it, 

strangely the results of expected bounding boxes were all the same (locations and sizes) among the test pictures.

I would like to know why.

Is it O.K. to change the name of first layer? 

Any comments are welcome!

**Comment by vgenty** [comment]
You should provide more details on training. Did the training loss go down (class loss, bounding box loss, rpn loss etc...) ? What is mAP at each epoch? Hard to debug without more information.

**Comment by LittleWat** [comment]
Thank you for replying. The loss didn't decrease.

I noticed that learning doesn't work well if i change the first layer name("conv1_1") .

I gave up to change the first layer.

**Comment by kexul** [comment]
@LittleWat Hi,did you change someplace elsewhere to make it work except the prototxt?I followed your way,but still encountered the same problem.

**Comment by zhenni** [comment]
@LittleWat @kexul Do you notice the learning rate of the first several layers are set as 0 in the examples? Do you change it to a normal value?

If you want to accept more channels, you have to change some code in `lib/roi_data_layer/minibatch.py` except the prototxt files.

**Comment by kexul** [comment]
@zhenni I've successfully modified the channels just as what you've mentioned.
I checked the learning rate setting,but found no clue that the learning rate of first several layers is 0. I didn't change the learning rate,the solving run smoothly and converged as expected.

**Comment by zhenni** [comment]
@kexul 
For example, this line in [pascal vgg16 ](https://github.com/rbgirshick/py-faster-rcnn/blob/master/models/pascal_voc/VGG16/faster_rcnn_end2end/train.prototxt#L21)

If you give more information about the training, for instance, the prototxt of training and where do you exactly change the code, maybe other people can help you. Do the result exactly the same? 

Oh, by the way, do you remove the pkl file everytime you make some changes for the dataset?

**Comment by lsj910128** [comment]
@kexul 
Hi, could you share some details about how to modified to channels?

**Comment by zhenni** [comment]
@kexul (@LittleWat )
For example, in `py-faster-rcnn/models/pascal_voc/VGG16/faster_rcnn_alt_opt/stage1_rpn_train.pt`.

```
layer {
  name: "conv1_1"
  type: "Convolution"
  bottom: "data"
  top: "conv1_1"
  param {
    lr_mult: 0
    decay_mult: 0
  }
  param {
    lr_mult: 0
    decay_mult: 0
  }
  convolution_param {
    num_output: 64
    pad: 1
    kernel_size: 3
  }
}
```
The learning rate has been set to be zero as `lr_mult: 0`.

If you change the number of channels, we you do copying from the vgg weights, I think there might be some error here. You may not copy the weights correctly as you expected. 
And if you just renaming the layer, the weights will be assigned to some random weights?

**Comment by wangwen39** [comment]
any other solutions for this issue? i try the above methods all but they didn't get any help........please help

**Comment by KevinHeish** [comment]
If you want to change the number of channels , you should change files below:

`"lib/utils/blob.py"` :
`blob = np.zeros((num_images, max_shape[0], max_shape[1], "Your channels") , dtype=np.float32)`

`"lib/roi_data_layer/layer.py"` :
`top[idx].reshape(cfg.TRAIN.IMS_PER_BATCH, "Your channels" , max(cfg.TRAIN.SCALES), 
  cfg.TRAIN.MAX_SIZE)`

I modified the `"lib/roi_data_layer/minibatch.py"` because I need to merge two image as input.
You should also change your `"test.prototxt"` and `"test.py"` when you run test.

**Comment by LittleWat** [comment]
I could solve this. Thank you all!

**Comment by ghost** [comment]
@LittleWat could you give details of what all changes you had to make to get this working?

I am seeing this 
```
F1023 14:37:54.407449 28982 net.cpp:774] Cannot copy param 0 weights from layer 'conv1_1'; shape mismatch.  Source param shape is 64 3 3 3 (1728); target param shape is 64 4 3 3 (2304). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.
```
after following the steps mentioned by @KevinHeish .
Do we have to change the name of the first layer in `train.prototxt`?
